wallet: rpc to add automatically generated descriptors

achow101 opened the pull request:
It is useful to have a RPC that can create and add the automatically generated descriptors (that are normally made during creation) to a wallet. This is especially useful when a new default descriptor has been implemented as it allows wallets created before that time to quickly and easily add that type of descriptor.

In particular, descriptor wallets created before Taproot was implemented can use the new `createwalletdescriptor` RPC in order to get a Taproot descriptor.

Furthermore, to keep the newly generated descriptor in line with the existing descriptors, this PR uses #26728 as it exposes a "wallet extended key" for us (in addition to upgrading wallets implemented prior to have a wallet xpub). The newly generated descriptors will use the "wallet extended key" stored in `CWallet` that PR adds.

`createwalletdescriptor` is generic and so it can also be used with the other address types. Of course, it given the same wallet extended key, address type, and internal-ness, it will create the same descriptor. So some refactoring has been done in order to detect that the same descriptor is about to be created, and to avoid overwriting any existing descriptors.

achow101 committed:
descriptor: Be able to get the pubkeys involved in a descriptor

achow101 committed:
wallet, rpc: Add createwalletdescriptor RPC

achow101 committed:
wallet, descspkm: Refactor wallet descriptor generation to static func

achow101 committed:
wallet: Test createwalletdescriptor

luke-jr commented:
I would expect `upgradewallet` to do this. Is there a reason for the approach here?

luke-jr commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,159 @@ RPCHelpMan simulaterawtransaction()
 
+static RPCHelpMan setdescriptorhdkey()
+{
+    return RPCHelpMan{"setdescriptorhdkey",
```
Any reason not to overload `sethdseed` for this?

Sjors replied:
Same question, couldn't `sethdseed` just have an option where it doesn't add fresh descriptors?

achow101 replied:
Changed to overload `sethdseed`.

Sjors replied:
Don't forget to update the PR description and the relevant error message in `createwalletdescriptor`.

achow101 commented:
> I would expect `upgradewallet` to do this. Is there a reason for the approach here?

The wallet version isn't modified by this. Also this allows for much more granulated control of what is added. While this can be used for upgrading a wallet to support a new kind of descriptor, it's not strictly just an upgrade function as it can also be used on blank wallets and after rotating the hd key. So I didn't think it was appropriate for `upgradewallet`.

w0xlt commented:
Concept ACK

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -754,6 +754,159 @@ static RPCHelpMan migratewallet()
+        "The address type must be one that the wallet does not already have a descriptor for."
+        + HELP_REQUIRING_PASSPHRASE,
+        {
+            {"address_type", RPCArg::Type::STR, RPCArg::Optional::NO, "The address type the descriptor will produce"},
```
Maybe list the address types here.

Sjors reviewed (commented):
Concept ACK on `createwalletdescriptor`.

Sjors commented:
The GUI receive address type dropdown is not updated after calling this.

Sjors commented:
Suggested testing for other reviewers:
1. Create a blank descriptor wallet
2. Call `sethdseed false` to create a seed without any descriptors (check with `listdescriptors`)
3. Call `createwalletdescriptor bech32m`

> The GUI receive address type dropdown is not updated after calling this.

This is still not working (I have to close and reload the wallet).

Will review code soon(tm).

achow101 commented:
> > The GUI receive address type dropdown is not updated after calling this.
> 
> This is still not working (I have to close and reload the wallet).

I think fixing that is out of scope for this PR. It's generally an issue for adding descriptors to a wallet.

Rspigler commented:
tACK 6db7d39efef93df728efa4c592b5e63d80409e80

> Create a blank descriptor wallet

:heavy_check_mark: 

> Call sethdseed false to create a seed without any descriptors (check with listdescriptors)

`sethdseed`

> null

`listdescriptors`

> {
>   "wallet_name": "pr25907",
>   "descriptors": [
>   ]
> }

`createwalletdescriptor bech32m`

> 
> {
>   "descs": [
>     "tr([4ad34cda/86'/0'/0']xpub6BxFBZ3fny6ijY6CUp2zudag83XqefXZZa7CdzxEx6K9kbLg9jLEArhWNaxtWsZ5wRPiAU6mJTDnUgt6uLLzfqyDHudbyjvZPBNBLSvwxhG/0/*)#x60zmzje",
>     "tr([4ad34cda/86'/0'/0']xpub6BxFBZ3fny6ijY6CUp2zudag83XqefXZZa7CdzxEx6K9kbLg9jLEArhWNaxtWsZ5wRPiAU6mJTDnUgt6uLLzfqyDHudbyjvZPBNBLSvwxhG/1/*)#hw2rxhzp"
>   ]
> }

Sjors reviewed (commented):
5e749f1d5bd8409f4fa1a5073c0340c68e12092c causes a test failure:

```
2022-10-05T18:07:50.143000Z TestFramework (INFO): Test disabled RPCs
2022-10-05T18:07:50.189000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_test()
  File "test/functional/wallet_descriptor.py", line 104, in run_test
    assert_raises_rpc_error(-4, "Only legacy wallets are supported by this command", recv_wrpc.rpc.sethdseed)
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/util.py", line 130, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised
2022-10-05T18:07:50.247000Z TestFramework (INFO): Stopping nodes
```

It would also be a bit easier to review if you can do the whitespace / documentation cleanup separate from the functionality change. Right now I find it easier to just re-review the whole `sethdseed` function, though that's small enough, so I don't mind.

The other commits look mostly fine to me (as of 6db7d39efef93df728efa4c592b5e63d80409e80).

Sjors commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -823,6 +823,102 @@ static RPCHelpMan migratewallet()
+        + HELP_REQUIRING_PASSPHRASE,
+        {
+            {"address_type", RPCArg::Type::STR, RPCArg::Optional::NO, "The address type the descriptor will produce. Options are \"legacy\", \"p2sh-segwit\", \"bech32\", and \"bech32m\"."},
+            {"internal", RPCArg::Type::BOOL, RPCArg::DefaultHint{"Both external and internal will be generated unless this parameter is specified"}, "Whether to only make one descriptor that is internal (if parameter is true) or external (if parameter is false)"},
```
d13ebb5a43733a61245cdf396304c3ff67b64900: slight preference to drop this. It's confusing and I personally don't see any use for it. For advanced use case it seems better to me if users to manually construct the descriptor (with help from `getxpub` #22341). But maybe I missed something.

achow101 replied:
I am ambivalent but I think it is a useful distinction as it can be used to rotate only specific descriptors rather than complete sets.

Sjors commented on test/functional/wallet_descriptor.py:
```diff
@@ -101,7 +105,6 @@ def run_test(self):
         assert_raises_rpc_error(-4, "Only legacy wallets are supported by this command", recv_wrpc.rpc.dumpprivkey, recv_wrpc.getnewaddress())
         assert_raises_rpc_error(-4, "Only legacy wallets are supported by this command", recv_wrpc.rpc.dumpwallet, 'wallet.dump')
         assert_raises_rpc_error(-4, "Only legacy wallets are supported by this command", recv_wrpc.rpc.importwallet, 'wallet.dump')
-        assert_raises_rpc_error(-4, "Only legacy wallets are supported by this command", recv_wrpc.rpc.sethdseed)
```
6db7d39efef93df728efa4c592b5e63d80409e80: I guess this was supposed to go in 5e749f1d5bd8409f4fa1a5073c0340c68e12092c

achow101 replied:
Oops, fixed.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -458,71 +458,140 @@ static RPCHelpMan unloadwallet()
-                "HD will have a new HD seed set so that new keys added to the keypool will be derived from this new seed.\n"
-                "\nNote that you will need to MAKE A NEW BACKUP of your wallet after setting the HD wallet seed." +
+        "Set or generate a new HD wallet seed or key.\n"
+        "\nLegacy wallets can only have a seed set. Non-HD Legacy wallets will not be upgraded to being a HD wallet."
```
No, Legacy wallets can only take a HD seed, not a key. This refers to the "or key" of the sentence above.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -458,71 +458,140 @@ static RPCHelpMan unloadwallet()
+        "Set or generate a new HD wallet seed or key.\n"
+        "\nLegacy wallets can only have a seed set. Non-HD Legacy wallets will not be upgraded to being a HD wallet."
+        "Legacy wallets that are already HD will have a new HD seed set so that new keys added to the keypool will be derived from this new seed.\n"
+        "\nDescriptor wallets can have either a HD seed or a HD key set. The seed or key will only be used for new automatically generated descriptors that are created when newkeypool is true, and with `createwalletdescriptors`.\n"
```
The RPC can accept a "HD key" for descriptor wallets.

S3RK commented:
I have a question related to `sethdseed`. I can see why we would want to be able to _set_ a seed for an empty wallet, this is use case is clear to me. However I don't fully understand yet the use case for being able to _change_ the seed for existing wallet. Is this intended? If yes, can we better describe the use case for this?

My best guess is that changing seed is needed for key rotation, but I think creating new wallet might actually be a better way to rotate keys. Separate wallets provide much better control over your identities by isolating them and avoiding accidentally mixing them up.

achow101 committed:
walletdb: Read and write activehdkey record

achow101 committed:
wallet: Load activehdkey and its private keys

The activehdkey record will now be loaded into the wallet as its own
member variable. Since the private keys will always be in existing
descriptors in the wallet, the key for the active hd key is also pulled
out of those descriptors on loading.

achow101 committed:
wallet: Store master key used for automatically generated descriptors

achow101 committed:
wallet: Retrieve active hd pubkey

achow101 committed:
wallet, rpc: Add gethdkey RPC

achow101 committed:
wallet: Implement GetActiveHDPrivkey

achow101 committed:
key: Add constructor for CExtKey that takes CExtPubKey and CKey

We often need to construct a CExtKey given an CExtPubKey and CKey, so
implement a constructor that does that for us.

achow101 committed:
wallet: Automatically upgrade a wallet to have global hd key

achow101 committed:
test: Test automatic upgrade of descriptor wallets to have hd key

achow101 committed:
wallet: Refactor function for single DescSPKM setup

We will need access to a function that sets up a singular
DescriptorSPKM, so refactor this out of the multiple DescriptorSPKM
setup function.

achow101 commented:
I've rebased/reimplemented this on top of #26728. The behavior that this PR adds doesn't change, just how it actually does it under the hood.

achow101 committed:
tests: Test for gethdkey

achow101 commented:
Per the discussion during the IRC meeting on 2023-01-27, I've removed the ability to rotate the active hd key. `sethdseed` can be used to set one on a blank wallet, but if there is already on set, then it cannot be changed.

achow101 committed:
wallet: Always set WALLET_FLAG_GLOBAL_HD_KEY for new wallets

achow101 committed:
wallet: Set global hd key for migrated wallets

achow101 committed:
wallet, rpc: Check WALLET_FLAG_GLOBAL_HD_KEY in gethdkey

achow101 committed:
test: Also do backwards compat test with encrypted wallets

Best reviewed with `git show -w`

achow101 committed:
test: Check that no unencrypted records persist after encrypting

achow101 commented:
#26728 was changed to not be writing the `hdkey` and `hdckey` records for backwards compatibility reasons. This PR has added those back in, but with a required wallet flag of `WALLET_FLAG_HAS_HDKEY_RECORDS` to prevent the downgrade issue described in https://github.com/bitcoin/bitcoin/pull/26728#issuecomment-1826396081

This flag will only be set when `sethdseed` is used with a descriptor wallet. `createwalletdescriptor` will should still work without those records and that flag.

Sjors commented:
This new approach makes sense to me. However, IIUC this PR doesn't require the new `sethdseed` RPC, so maybe move that to its own PR? (personally I care a lot less about seed rotation than about being able to add taproot descriptors to an existing descriptor wallet)

achow101 commented:
> However, IIUC this PR doesn't require the new `sethdseed` RPC, so maybe move that to its own PR?

It's mainly there for the test to work.

furszy committed:
test: add coverage for re-opening a downgraded encrypted wallet on master

The test creates a wallet on master, downgrades and encrypts the wallet.
Then, it tries to open it again on master.

Sjors commented:
You could also test `createwalletdescriptor` by creating a descriptor wallet using a pre-taproot version. You can check the xpub in the resulting descriptor using #22341. That way all the key rotation stuff and the new records can be done in another PR.

achow101 committed:
walletdb: Check that unencrypted and crypted keys are mutually exclusive

achow101 commented:
I've split the `sethdseed` changes into #29054

achow101 commented:
Superseded by #29130

I don't think there's a reason to come back to this design, but the branch won't be deleted so we can reopen if desired.
